### DISCUSSION TITLE
iridium run command prints "Todo" and iridium init command is giving error

### DISCUSSION THREAD

**Discussion by erzohrakhan** [IridiumIdentity/iridium#100] [Q&A]
Hi,
I have tried  to run Iridium locally but I am facing problem:
https://docs.iridium.software/0.1.4/3.-for-developers/building-from-source/
1. `mvn clean package `  with this command build is successful.
2. `{ProjectRoot}/iridium-cli/target/iridium-{iridium-version}-bin.tar.gz` ; extracted the files from this tar and tried to irdium(under bin folder); I got this error 
```
bin % ./iridium init
16:29:57.210 [main] INFO  s.i.cli.generator.PersistencePropertyGenerator - creating persistence properties...
16:29:57.326 [main] WARN  o.h.j.b.internal.EntityManagerFactoryBuilderImpl - HHH000059: Defining hibernate.transaction.flush_before_completion=true ignored in HEM
16:29:57.434 [main] WARN  org.hibernate.orm.connections.pooling - HHH10001002: Using built-in connection pool (not intended for production use)
16:29:57.505 [main] WARN  o.h.e.jdbc.env.internal.JdbcEnvironmentInitiator - HHH000342: Could not obtain connection to query metadata
java.lang.IllegalStateException: Cannot get a connection as the driver manager is not properly initialized
        at org.hibernate.engine.jdbc.connections.internal.DriverManagerConnectionProviderImpl.getConnection(DriverManagerConnectionProviderImpl.java:262)
        at org.hibernate.engine.jdbc.env.internal.JdbcEnvironmentInitiator$ConnectionProviderJdbcConnectionAccess.obtainConnection(JdbcEnvironmentInitiator.java:316
```) 

and iridium run command is printing "TodO"
Let me know what I am missing

**Comment by joshfischer1108** [comment]
Thanks for your question, @erzohrakhan.  We've been rapidly developing on `main` and working towards a new release.  The upcoming release will fix the issue you are experiencing.  

If you want to use iridium before the next release, you can clone the repo and build off `main`, following the instructions you listed in steps `1` and `2`.

It would be easiest (yet still tedious at the moment :-)) for you to run the API from your IDE after you init the database.   the [docs](https://docs.iridium.software/NEXT/) for the `Next` version may be a little more helpful than `0.1.4`.   Of course, they are still a work in progress so any questions, or recommendations from you on how to make the system easier to use and understand would be greatly appreciated.   


One thing, before you run `iridium init`.  With the `0.2.0-SNAPSHOT` changes all arguments have been removed from the command and are driven by yaml files in the `conf` folder  of the release `.tar.gz` file you generate after building from source.  

So it should be something like

```
$ mvn clean package
$ tar -xzf iridium-cli/target/iridium-0.2.0-SNAPSHOT-bin.tar.gz
$ vim iridium-0.2.0-SNAPSHOT-bin/conf/external-providers.yaml
```

The configurations you place in this file, determine the social providers that will be available for login to the system. You can add or remove github or google as you see fit. 

I understand the documentation is lacking at the moment.   I'd be happy to help with any other questions you have.

**Reply by erzohrakhan** [reply]
I would love to contribute to this project

**Reply by joshfischer1108** [reply]
@erzohrakhan we'd appreciate your help!

**Comment by joshfischer1108** [comment]
I also noticed the error message
```
16:29:57.505 [main] WARN  o.h.e.jdbc.env.internal.JdbcEnvironmentInitiator - HHH000342: Could not obtain connection to query metadata
java.lang.IllegalStateException: Cannot get a connection as the driver manager is not properly initialized
```
This suggests the CLI tool could not get a connection to your database.  Did you check the credentials supplied are correct?

**Comment by erzohrakhan** [comment]
I have given the following commands to set up the database
```
$ docker run  --name iridium-db \
    -p 3306:3306 \
    -e MYSQL_ROOT_PASSWORD=iridium \
    -e MYSQL_DATABASE=identities \
    -e MYSQL_USER=identity-api \
    -e MYSQL_PASSWORD=iridium \
    -d mariadb:latest

$ docker start iridium-db
```

But now I am getting this error:
```
00:50:07.047 [main] WARN  org.hibernate.engine.jdbc.spi.SqlExceptionHelper - SQL Error: 1045, SQLState: 28000
00:50:07.048 [main] ERROR org.hibernate.engine.jdbc.spi.SqlExceptionHelper - Access denied for user 'iridium'@'172.17.0.1' (using password: YES)
```

**Reply by rwithers** [reply]
Hi @erzohrakhan There are a couple things that I would recommend here that might help.  The first one would be just run the system don't try to build from source.  The latest coming right of the main branch along with the nightly builds in docker should just about get you where you want to go.  I would also start everything with the "defaults" at first and then modify the values to your liking once you've hit the redirect with one of (github / google) providers. 

1. Pull latest from main 
2. Make sure the docker compose file is pointing at iridiumidentity/iridium-core-server-nightly:latest
3. run `mvn clean install`
4. run ./bootIridium.sh Please Note: the bootIridium.sh script makes assumptions about what you have installed currently. 
5. The bootIridium.sh script will complain if you don't have a clientid or secret in the external-providers.yaml file.  
6. You can read about this process here: 
https://docs.github.com/en/apps/oauth-apps/building-oauth-apps/authenticating-to-the-rest-api-with-an-oauth-app
and 
You can register here: https://github.com/settings/applications/new

Now open a new terminal once you see the ui come up.  In the new terminal run the init script.  
This can be done with the command `iridium init`

I would follow the process above to get the whole system working, then work on pulling it apart in its various components.

I hope this helps.

**Comment by joshfischer1108** [comment]
@erzohrakhan We pushed a new release.  Let me know how the below docs work out for you.

https://docs.iridium.software/0.2.0/2.-getting-started/2.run-iridium-locally/

**Reply by erzohrakhan** [reply]
I'm getting the following error for the command 
`java -jar ./lib/iridium-core-server-0.2.0.jar --spring.config.location=../conf/application.properties
`
***************************
APPLICATION FAILED TO START
***************************

Description:

Failed to configure a DataSource: 'url' attribute is not specified and no embedded datasource could be configured.

Reason: Failed to determine a suitable driver class


Action:

Consider the following:
	If you want an embedded database (H2, HSQL or Derby), please put it on the classpath.
	If you have database settings to be loaded from a particular profile you may need to activate it (no profiles are currently active).

**Comment by joshfischer1108** [comment]
Good catch, @erzohrakhan .  1. I had a typo in the properties file path.  2.  It seems the jar isn't honoring the command line args that are passed to it.

A way to get around this is to create an environment variable to define where the `application.properties` file is.  Assuming you are in the top-level folder of the distribution you downloaded. Try the following.

```bash
$ export SPRING_CONFIG_LOCATION=./conf/application.properties
$  java -jar ./lib/iridium-core-server-0.2.0.jar
```

**Comment by erzohrakhan** [comment]
Thanks, Josh
It worked

**Reply by joshfischer1108** [reply]
I've since updated the docs on 0.2.0 as well. Thanks for reporting the issue, 💯 .
